use of strsep() fails to find function definition in strings.h on Windows with clang

I get this warning:

D:/a/NCEPLIBS-g2c/NCEPLIBS-g2c/g2c/src/g2ccsv.c:222:23: error: implicit declaration of function 'strsep'; did you mean 'strset'? [-Wimplicit-function-declaration]
  222 |                 tmp = strsep(&buf, "\"");
      |                       ^~~~~~
      |                       strset
D:/a/NCEPLIBS-g2c/NCEPLIBS-g2c/g2c/src/g2ccsv.c:222:21: error: assignment to 'char *' from 'int' makes pointer from integer without a cast [-Wint-conversion]
  222 |                 tmp = strsep(&buf, "\"");
      |                     ^
D:/a/NCEPLIBS-g2c/NCEPLIBS-g2c/g2c/src/g2ccsv.c:223:21: error: assignment to 'char *' from 'int' makes pointer from integer without a cast [-Wint-conversion]
  223 |                 tmp = strsep(&buf, "\"");
      |                     ^
D:/a/NCEPLIBS-g2c/NCEPLIBS-g2c/g2c/src/g2ccsv.c:225:21: error: assignment to 'char *' from 'int' makes pointer from integer without a cast [-Wint-conversion]
  225 |                 tmp = strsep(&buf, ",");
      |                     ^
D:/a/NCEPLIBS-g2c/NCEPLIBS-g2c/g2c/src/g2ccsv.c:229:21: error: assignment to 'char *' from 'int' makes pointer from integer without a cast [-Wint-conversion]
  229 |                 tmp = strsep(&buf, ",");
      |                     ^

Looks like we should use strtok() instead. See https://stackoverflow.com/questions/58244300/getting-the-error-undefined-reference-to-strsep-with-clang-and-mingw.

@AlysonStahl-NOAA can you convert this code to strtok()? Hopefully sometime this week?

--- a/src/g2ccsv.c
+++ b/src/g2ccsv.c
@@ -188,7 +188,7 @@ g2c_find_entry(char *desc, G2C_CODE_TABLE_T *table)
  *
  * @return a pointer to the original value of stringp
  */
-char *
+static char *
 g2c_csv_strsep(char **stringp, const char *delim)
 {
     char *rv = *stringp;

--- a/src/g2ccsv.c
+++ b/src/g2ccsv.c
@@ -177,7 +177,6 @@ g2c_find_entry(char *desc, G2C_CODE_TABLE_T *table)
     return NULL;
 }
 
-
 /** Implementation of strsep for code portability.
  * Extracts first token in string given a delimiter.
  *

--- a/src/g2ccsv.c
+++ b/src/g2ccsv.c
@@ -189,7 +189,8 @@ g2c_find_entry(char *desc, G2C_CODE_TABLE_T *table)
  *
  * @return a pointer to the original value of stringp
  */
-char* g2c_csv_strsep(char **stringp, const char *delim)
+char *
+g2c_csv_strsep(char **stringp, const char *delim)
 {
     char *rv = *stringp;
     if (rv)

--- a/src/g2ccsv.c
+++ b/src/g2ccsv.c
@@ -177,6 +177,32 @@ g2c_find_entry(char *desc, G2C_CODE_TABLE_T *table)
     return NULL;
 }
 
+
+/** Implementation of strsep for code portability.
+ * Extracts first token in string given a delimiter.
+ *
+ * @param stringp The address of a pointer to string to be separated.
+ * This value is overwritten with the value past the delimiter.
+ * @param delim Characters that delimit the tokens.
+ *
+ * @author Alyson Stahl @date 11/6/24
+ *
+ * @return a pointer to the original value of stringp
+ */
+char* g2c_csv_strsep(char **stringp, const char *delim)
+{
+    char *rv = *stringp;
+    if (rv)
+    {
+        *stringp += strcspn(*stringp, delim);
+        if (**stringp)
+            *(*stringp)++ = '\0';
+        else
+            *stringp = 0;
+    }
+    return rv;
+}
+
 /**
  * Initialize tables from "CodeFlag.txt".
  *
@@ -219,14 +245,14 @@ g2c_csv_init()
 
             if (*buf == '\"')
             {
-                tmp = strsep(&buf, "\"");
-                tmp = strsep(&buf, "\"");
+                tmp = g2c_csv_strsep(&buf, "\"");
+                tmp = g2c_csv_strsep(&buf, "\"");
                 key = strdup((const char *)tmp);
-                tmp = strsep(&buf, ",");
+                tmp = g2c_csv_strsep(&buf, ",");
             }
             else
             {
-                tmp = strsep(&buf, ",");
+                tmp = g2c_csv_strsep(&buf, ",");
                 key = strdup((const char *)tmp);
             }
 

